rfcs/react-components/convergence/long-term-support-plan.md

```diff
@@ -0,0 +1,122 @@
+
+We should add pointers from the wiki FAQ back to the README.MD for each version.
+
+#### Retire v7
+
+We should announce an end-of-life to v7. Proposed: June 30th, 2024. We would lock the branch from any changes, disable and archive any build pipelines, and replace public documentation sites with a redirect to a notice that v7 is retired.
+
+Any partners wanting to continue using v7, should take a snapshot of the code and integrate it into their own repositories and build it.
+
+#### Separate v0, v8 and v9 branches
+
+v7 is already in its own branch. v0, v8, and v9 are in the master branch which causes some confusion for developers.
+
+1. Create a v8 branch from master
+2. Create a northstar-v0 branch from master
+3. Remove all v9 code from the northstar-v0, and v8 branches
+4. Remove all v0 and v8 code from the master branch
+5. Update v8 components to take npm dependencies on compat components
+6. Update any migration components to take npm dependencies on v8 components
+7. Update v0, v8 builds to publish only v0, v8 components
+8. Update v9 build to publish only v9 components
+9. Update README.MD in each branch to cover its version and point to the other branches
+10. Update ADO pipelines to handle each different branch (triggers, builds, releases)
+11. Update ADO documentation for releasing each version
+
+- Keep cross-version compat components in master (v9).
+- Keep migration components in master (v9).
+
+This plan requires communicating to partners that own their own v0 and v8 based packages in the master branch that they will need to start working on them in the new branches.
+
+There should be a cut-over date where branches are locked for changes and then the steps taken (scripts?) to make the change followed by the branches being unlocked. It would be good to time
+the cutover to a time of low activity for all teams globally.
+
+#### Flatten master branch
+
+With northstar-v0 and v8 no longer in the master branch, we can reduce the depth of the tree by bringing components within the react-components folder up one level.
+
+#### Sunset v8
+
+After retiring v7, announce v8 is in sunset mode. Only critical bugs will be fixed. We will no longer be moving React forward in v8. Set a retirement date for one year out.
+
+#### Define team tenets for active versions
+
+The Fluent team size doesn't allow for supporting more than 2 versions of Fluent. We should consider policy where we don't start active development on a new major version until the oldest version is retired (or at least in sunset). We can have a next branch and do lots of experimentation there, but there would be no support for this branch.
+
+Branch renames make it possible to have an easier current verson change if we do next version development it its own branch. We can rename master to vX and then rename next to master.
+
+Note: We should also consider renaming master to main.
+
+### Pros and Cons
```

Comment:
## I'll recap my overall POV/suggested approach here:

In general, branching old versions is common pattern in OSS and I fully support this approach, but I'd not apply it for us at this stage. 

Having live(maintenance) branches in a monorepo environment comes with lot of overhead (package bumps/security/tooling tweaks/etc). 

While we could postpone all of this until the tooling work is done (migration to nx etc, having automated tooling migrations in place that could be used in branches), that would still not give us any future guarantees. 

In general any react implementations that are not declared as EOL should stay in master + re-structuring our folder structure seems to be enough.

if we introduced proposed branches now, any propagating changes back and forth would add additional churn ( release from branch a, bump in branch b and repeat )

### v7:
- EOL / block and freeze - 👍

### Branching v8 and v0:

**build times + builds more reliable**
  - yes this would "speed" up things "instantly", but that's not a long term solution as we need improve our whole CI situation, because as we are adding more code to ACTIVE we can/will end up in the same situation as we have today in upcoming months/years. 
  - we are not using cached builds, once we will have this in place  build times for changes that need v0,v8 assets as part of affected task tree will no longer be a concern.
  - master pipeline is always running everything, which doesn't make any sense and consumes lot of resources. With cache and running effective affected trees since last master green run will drastically improve the master pipeline situation.
  - more details with all the data and actionable are well documented here https://github.com/microsoft/fluentui/issues/27758

**folder depth is reduced and packages more discoverable**
 
We don't need branches to improve this. Also what will happen to `web-components` (any kind of framework/platform agnostic living in our monorepo will need to follow our patterns) ?

**I'd suggest following approach:**

re-shuffling current folder structure to make it more "accessible", organized by "domain/framework"

master branch:

```sh
/
 |- react/  # ACTIVE fluent react implementation (v9)
 |- react-v8/  # MAINTENANCE fluent react implementation (v8)
 |- react-v0/  # MAINTENANCE fluent react implementation (v0)
 |- web-components/
 |- shared/ # any shared logic domain between implementation (tokens for instance)
 |- tools/ # general purpose repo infra/tooling/etc
 |- package.json
 |- nx.json
```

which after declaring v8 and v0 as EOL would move to

master branch:

```diff
/
 |- react/  
- |- react-v8/  
- |- react-v0/ 
 |- web-components/
 |- shared/
 |- tools/
 |- package.json
 |- nx.json
```

v8 branch:
```diff
/
 |- react/ # 'react-v8' folder renamed to `react` to mirror our implementation per branch pattern
 |- tools/
 |- package.json
 |- nx.json
```

v0 branch:
```diff
/
 |- react/ # 'react-v8' folder renamed to `react` to mirror our implementation per branch pattern
 |- tools/
 |- package.json
 |- nx.json
```

Now after (v0 and v8 EOL) we will follow same branching pattern for releases after v9 ( maybe we can already branch MAINTENANCE  in its `v.x` branch ).

Reply:
IF the general agreement would be to branch right away from implementation POV we would need implement:

- move project generator 
- remove project generator
- creating new pipelines/redoing existing pipelines
- implementing automatic migrations from master to MAINTENANCE branches

Reply:
We should chat in person more about not moving to a branch-per-version strategy. I hadn't considered web-components in this plan.  I guess if we don't think of our repo holding one thing (the Fluent UI React component library) then it doesn't really have a single version. I'll set up a time to chat this week.

Reply:
I've updated the doc to address what we discussed. :-)